font/src/ft/mod.rs

```diff
@@ -537,48 +607,66 @@ impl FreeTypeRasterizer {
-                            // Safe to unwrap the option since we've already checked for the path
-                            // and index above.
-                            let key = self.face_from_pattern(&pattern)?.unwrap();
-                            Ok(key)
-                        },
+        // This check should be pretty fast and save us from traversing the entire list
+        if !fallback_list.coverage.has_char(ch) {
+            return Ok(glyph.font_key);
+        }
+
+        for font_pattern in &fallback_list.list {
+            let path = match font_pattern.file(0) {
+                Some(path) => path,
+                None => continue,
+            };
+
+            match self.keys.get(&path) {
+                Some(&key) => {
+                    let face = match self.faces.get(&key) {
+                        Some(face) => face,
+                        None => continue,
+                    };
+
+                    let index = face.ft_face.get_char_index(ch as usize);
+
+                    // We found something in a current face, so let's use it
+                    if index != 0 {
+                        return Ok(key);
                     }
-                } else {
-                    Err(Error::MissingFont(FontDesc::new(
-                        "fallback-without-path",
-                        Style::Specific(glyph.to_string()),
-                    )))
-                }
-            },
-            None => Err(Error::MissingFont(FontDesc::new(
-                "no-fallback-for",
-                Style::Specific(glyph.to_string()),
-            ))),
+                },
+                None => {
+                    // This path will be hit, only if you don't have glyph at all or didn't
+                    // loaded font, so not that hot for a normal scenario and all costly parts
+                    // are inside the if
+                    let pattern = font_pattern.clone();
+                    let charset = pattern.get_charset();
+
+                    if charset.has_char(ch) {
+                        let config = fc::Config::get_current();
+
+                        // Recreate a pattern
+                        let mut pattern = fc::Pattern::new();
+                        pattern.add_pixelsize(self.pixel_size as f64);
+                        // Unwrap should be safe in general, however let's play safe here
+                        pattern.add_style(font_pattern.style().next().unwrap_or_else(|| "Regular"));
+                        pattern.add_family(
+                            font_pattern.family().next().unwrap_or_else(|| "monospace"),
+                        );
+                        // We should render pattern, otherwise most of its properties wont work
+                        let pattern = pattern.render_prepare(config, font_pattern);
+
+                        let key = self.face_from_pattern(&pattern, None)?.unwrap();
```

Comment:
Unwrapping on this seems strange? How do we know it will always be `Some`?

Reply:
You can get `None` if your pattern doesn't have path, but you're already matching by path in this loop, so you know that pattern has path, that means that check can't fail.